### PR TITLE
core: announce ChainSideEvent during reorg

### LINKED ISSUE #2298: Block is lost and no one wrapped it as uncle, even myself
I found block at height N. Usually I expect it to be at least an uncle in case of race. But no one included it as uncle in next 7 blocks. Also I found N+5 block and if my block is in the same geth instance block N was not included as uncle in N+5 block. This is weird. Can anyone explain how that happened? I have more than 96 peers for propagation.

This is not a single case, day before that happened too and I remember some lost blocks since ETH launch. Is there unresolved issues in network protocol? If I can't include my own block as uncle in my own block I think it's a design flaw or at least a bug.

### PR DESCRIPTION
**Pull request by obscuren**
Previously all blocks that were already in our chain were never re
announced as potential uncle block (e.g. ChainSideEvent). This is
problematic during mining where you want to gather as much possible
uncles as possible increasing the profit. This is now addressed in this
PR where during reorganisations of chains the old chain is regarded as
uncles.

Fixed #2298

### CODE REVIEW

**Comment by robotally** [comment]
| Vote | Count | Reviewers |
| :---: | :---: | :---: |
| :+1: | 0 |  |
| :-1: | 0 |  |

_Updated: Mon Mar 7 17:12:53 UTC 2016_

**Comment by karalabe** [comment]
Hmm, all of the CI servers failed with some bad block error.

**Comment by karalabe** [comment]
LGTM :+1:

File: core/blockchain.go [hunk 1]
```diff
@@ -1369,6 +1376,15 @@ func (self *BlockChain) reorg(oldBlock, newBlock *types.Block) error {
 		go self.eventMux.Post(RemovedLogsEvent{deletedLogs})
 	}
 
+	if len(oldChain) > 0 {
+		fmt.Println("side chain length =", len(oldChain))
```
Comment (fjl) [inline core/blockchain.go]: `glog.Printf`

Comment (karalabe) [inline core/blockchain.go]: Ah, I think that was a debug log for Jeff to find the test issue. Wasn't in the code before :D :P

File: core/blockchain.go [hunk 2]
```diff
@@ -1313,6 +1318,8 @@ func (self *BlockChain) reorg(oldBlock, newBlock *types.Block) error {
 			commonBlock = oldBlock
 			break
 		}
+
+		oldChain = append(oldChain, oldBlock)
 		newChain = append(newChain, newBlock)
 		deletedTxs = append(deletedTxs, oldBlock.Transactions()...)
```
Comment (fjl) [inline core/blockchain.go]: Can you explain why the tracking is also happening for the common block here?

Comment (karalabe) [inline core/blockchain.go]: That's not the common block, but the common height. If you reach the common block (lines above) you break out.

File: core/blockchain.go [hunk 3]
```diff
@@ -1369,6 +1376,15 @@ func (self *BlockChain) reorg(oldBlock, newBlock *types.Block) error {
 		go self.eventMux.Post(RemovedLogsEvent{deletedLogs})
 	}
 
+	if len(oldChain) > 0 {
+		fmt.Println("side chain length =", len(oldChain))
+		go func() {
+			for _, block := range oldChain {
+				self.eventMux.Post(ChainSideEvent{Block: block, Logs: deletedLogsByHash[block.Hash()]})
+			}
+		}()
```
Comment (fjl) [inline core/blockchain.go]: Can you consider moving the other `Post`s into this goroutine?

File: core/blockchain.go (hunk 1 above)
Comment (obscuren) [inline core/blockchain.go]: Oops  :-D

File: core/blockchain.go (hunk 3 above)
Comment (obscuren) [inline core/blockchain.go]: Grouping of the goroutines and possibly the `len` check is done in another PR. Want to keep the changes to a minimum.

File: core/blockchain.go (hunk 2 above)
Comment (fjl) [inline core/blockchain.go]: Aww, ok, I got confused by GH diff view.
